### PR TITLE
corecfg: allow using `# snapd-edit: no` header to disable pi-config

### PR DESCRIPTION
**Pull request by mvo5**
On some systems the pi config.txt cannot be modified safely. E.g.
on devices that measure boot or similar.
    
This commit adds a way to disable the system.pi-config.* settings
when adding the following header to the config.txt

    # snapd-edit: no

based on https://github.com/snapcore/snapd/pull/10448

### CODE REVIEW

File: overlord/configstate/configcore/picfg.go
Comment (bboozzoo) [inline overlord/configstate/configcore/picfg.go:106]: hmm how about `configuration cannot be applied: %s` ?

File: overlord/configstate/configcore/picfg.go
Comment (bboozzoo) [inline overlord/configstate/configcore/picfg.go:165]: Maybe this should be a Noticef right now?

File: overlord/configstate/configcore/picfg.go
Comment (pedronis) [inline overlord/configstate/configcore/picfg.go:118]: I think we want to read the full first line and then apply a regexp to it so that we can ignore different spacing or case:

`# Snapd-Edit: no`

**Review by pedronis — COMMENTED** [review]
some suggestions

File: overlord/configstate/configcore/picfg.go
Comment (pedronis) [inline overlord/configstate/configcore/picfg.go:110]: what's the intention to allow random things at the end?

I would have thought we would do "\s*$"  maybe?

if you think we should allow random things,  we probably want at least a space after the "no", no?

Comment (mvo5) [reply overlord/configstate/configcore/picfg.go:110]: Thanks! Sorry, I think I misunderstood the earlier suggestion, I will only allow whitespace after the stanza.

File: overlord/configstate/configcore/picfg.go
Comment (bboozzoo) [inline overlord/configstate/configcore/picfg.go:126]: probably no need to convert to strings
```suggestion
	return reIgnorePrefix.Match(scanner.Bytes())
```

File: overlord/configstate/configcore/picfg.go
Comment (bboozzoo) [inline overlord/configstate/configcore/picfg.go:115]: perhaps it'd be useful to have a test which removes the file or makes it unreadable to make sure we don't do something silly

**Review by bboozzoo — APPROVED** [review]
LGTM

**Review by mardy — APPROVED** [review]
LGTM! Just please be careful when merging this: it would be nice if all references to the "SNAP-IGNORE" text were removed from commit messages, since they would just be misleading (given that the code uses "snapd-edit: no").

I recommend squashing and editing the commit message.

File: overlord/configstate/configcore/picfg.go
Comment (mvo5) [reply overlord/configstate/configcore/picfg.go:115]: Nice catch, it seems we are just silently doing nothing if there is no config.txt, no error and no new one is created. I think that is probably outside the scope of this PR as it seems we have this behavior since a long time. I guess it's also a bit of a question what the right behavior is in this case, probably error(?).

Comment (pedronis) [reply overlord/configstate/configcore/picfg.go:115]: remember that this code is called everywhere even on things that are not pis at all ATM, that's why I think it's like that

File: overlord/configstate/configcore/picfg.go
Comment (mvo5) [reply overlord/configstate/configcore/picfg.go:107]: I think I like this better, thank you!

File: overlord/configstate/configcore/picfg.go
Comment (mvo5) [reply overlord/configstate/configcore/picfg.go:171]: Indeed, I think that is now a better choice

**Comment by mvo5** [comment]
@pedronis is this something you want to review before the merge or are you happy with the existing reviews?

**Comment by pedronis** [comment]
@mvo this seems to be undoing my suggestions from the previous PR

File: overlord/configstate/configcore/picfg.go
Comment (pedronis) [inline overlord/configstate/configcore/picfg.go:157]: this seems to undo a change that was done in the previous PR

File: overlord/configstate/configcore/picfg.go
Comment (pedronis) [inline overlord/configstate/configcore/picfg.go:142]: same

**Review by pedronis — COMMENTED** [review]
this is strange

File: overlord/configstate/configcore/picfg.go
Comment (pedronis) [reply overlord/configstate/configcore/picfg.go:115]: I was confused, this is new code, in theory we could return an error if the error is not file does not exist, not sure it matters a lot in practice, as we will try to use the file again later and file again I suppose

File: overlord/configstate/configcore/picfg.go
Comment (pedronis) [inline overlord/configstate/configcore/picfg.go:162]: s/ignore/no-editing/ ?

**Review by pedronis — APPROVED** [review]
thanks, remark about the error now that we changed the header

File: overlord/configstate/configcore/picfg.go
Comment (mvo5) [reply overlord/configstate/configcore/picfg.go:115]: Yeah, I think so too, we will later read the file and fail with a error.

**Pull request merged**
